pkg/volume/cinder/attacher.go

```diff
@@ -0,0 +1,276 @@
+*/
+
+package cinder
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type cinderDiskAttacher struct {
+	host volume.VolumeHost
+}
+
+var _ volume.Attacher = &cinderDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &cinderPlugin{}
+
+const (
+	checkSleepDuration = time.Second
+)
+
+func (plugin *cinderPlugin) NewAttacher() (volume.Attacher, error) {
+	return &cinderDiskAttacher{host: plugin.host}, nil
+}
+
+func (plugin *cinderPlugin) GetDeviceName(spec *volume.Spec) (string, error) {
+	volumeSource, _ := getVolumeSource(spec)
+	if volumeSource == nil {
+		return "", fmt.Errorf("Spec does not reference a Cinder volume type")
+	}
+
+	return volumeSource.VolumeID, nil
+}
+
+func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, _ := getVolumeSource(spec)
+	volumeID := volumeSource.VolumeID
+
+	cloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+	instanceid, err := cloud.InstanceID()
+	if err != nil {
+		return err
+	}
+	attached, err := cloud.DiskIsAttached(volumeID, instanceid)
+	if err != nil {
+		// Log error and continue with attach
+		glog.Errorf(
+			"Error checking if volume (%q) is already attached to current node (%q). Will continue and try attach anyway. err=%v",
+			volumeID, instanceid, err)
+	}
+
+	if err == nil && attached {
+		// Volume is already attached to node.
+		glog.Infof("Attach operation is successful. volume %q is already attached to node %q.", volumeID, instanceid)
+		return nil
+	}
+
+	_, err = cloud.AttachDisk(instanceid, volumeID)
+	return err
+}
+
+func (attacher *cinderDiskAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
+	cloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return "", err
+	}
+	volumeSource, _ := getVolumeSource(spec)
+	volumeID := volumeSource.VolumeID
+	instanceid, err := cloud.InstanceID()
+	if err != nil {
+		return "", err
+	}
+	devicePath := ""
+	if d, err := cloud.GetAttachmentDiskPath(instanceid, volumeID); err == nil {
+		devicePath = d
+	} else {
+		glog.Errorf("%q GetAttachmentDiskPath (%q) gets error %v", instanceid, volumeID, err)
+	}
+
+	ticker := time.NewTicker(checkSleepDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		probeAttachedVolume()
+		select {
+		case <-ticker.C:
+			glog.V(5).Infof("Checking Cinder disk %q is attached.", volumeID)
+			if devicePath == "" {
+				if d, err := cloud.GetAttachmentDiskPath(instanceid, volumeID); err == nil {
+					devicePath = d
+				} else {
+					glog.Errorf("%q GetAttachmentDiskPath (%q) gets error %v", instanceid, volumeID, err)
+				}
+			}
+			if devicePath == "" {
+				glog.Errorf("Cinder disk (%q) is not attached yet", volumeID)
```

Comment:
Ditto: This is not an error. If you want to log it make it `glog.V(5).Infof` otherwise the logs will be filled with this.